lib/chore/duplicate_detector.rb

```diff
@@ -24,9 +25,11 @@ def initialize(opts={})
     # Duplicated messages will return true
     def found_duplicate?(msg_data)
```

Comment:
So there are 2 contexts where the dedupe logic is run:

1. Initial consume of the message
    - [dedupe key == SQS message id](https://github.com/Tapjoy/chore/blob/v4.3.0/lib/chore/queues/sqs/consumer.rb#L93-L95)
    - let's call this the _**consumer dedupe check**_
    - if a duplicate is found, the job code is not called the message is effectively _discarded by the consumer_

1. Worker begins work on the message

    - received from the consumer process
    - [dedupe key == `jobKlass.dedupe_lambda`](https://github.com/Tapjoy/chore/blob/v4.3.0/lib/chore/worker.rb#L45-L62)
    - let's call this the _**worker dedupe check**_
    - if a duplicate is found, the message is _deleted from SQS_

Both are called via [lib/chore/consumer.rb#L76-L78](https://github.com/Tapjoy/chore/blob/v4.3.0/lib/chore/consumer.rb#L69-L78).

It's the 2nd dedupe check that sometimes leads to too-early deletion of a message, and it is almost certainly caused by a delay between the consumer consuming a message and a worker being ready to pick up that message from the consumer. This delay could be arbitrarily high if all Chore workers are busy. 

Given the fact that what we are attempting to prevent is early deletion of a message, and early deletion can _only_ occur in the worker context, I think the timeout offset implementation should _only_ exist in the worker class.

@obrie What do you think?

Reply:
This makes sense to isolate the deduplication and SQS removal to worker only. I was also going to say something about tracking last runtime, that way we would know for sure if the message should be removed from the queue or not. I'm not sure if we're currently passing it around though, but could potentially add it to the cache

Reply:
@obrie It looks like we don't get the consume time data in a message's metadata, so we'd need to manage the implementation ourselves.

```ruby
m = q.receive_messages(max_number_of_messages: 1)
=> #<Aws::SQS::Message::Collection:0x0000000106ebd4b0
 @batches=
  [[#<Aws::SQS::Message:0x0000000106ebd730
     @client=#<Aws::SQS::Client>,
     @data=
      #<struct Aws::SQS::Types::Message
       message_id="fa6ce081-b18a-43e3-b543-ea18ff1fbf82",
       receipt_handle=
        "AQEBSHcv1W+f5pvWzqPcKxqNrAvCpZJp3dY58Gzdi2Eq9zhJj01z+l/ArgXK7DW2M42vbRzfw9/Xz+oHLz69TLxrzifNlIHcvjhVGHtz0jQ/K1RZu0+6RjcQ1FABVy7rs1l4dHjRYBVapuL0gXZ/efO8f6+xxKDJpx1K1QPEcWeuooAStq0dTAtlbWmY6M1tESv9Iqb2cmk7HQgpUAtKovvA+LTTqERvPC2jh9JrRLns+L6JI9Q39xZMldSdjHjdfXfYWKpJlGX5QxDjcSrSpYszXMXuOTXayUXsEcckcISTULMbOJPvC78GsOAP/MpHEVCQMXdtfHZUaoqL0n3ZPxIXsk1Nj9E0NhadU0Qr0F1lhZ/2nN6irEj/mZqfy2rpZL4mUlEcQuIVQEqcAi9zGpZ2OA==",
       md5_of_body="89763ace9d7c2249f482b20f3a0b454f",
       body=
        "{\"class\":\"NonManagedSendCurrencyJob\",\"args\":[\"d1e86a2d-1917-42fa-9afe-f5b4fff1d7e8\",{\"click_key\":\"tjid.6c332bb2c9844049801cc48c07c6e64a.e4881b5a-f839-4741-9f1b-9339e29e42a5\",\"install_timestamp\":\"1715710447.2181976\",\"options\":{\"device_identifier\":{\"id\":\"bc6acf602d0b401bb619ed21e3227d54\",\"type\":\"Advertising ID\"},\"source\":\"ConnectController\",\"conversion_ip\":\"197.43.188.132\",\"event_cost_id\":null,\"attribution_source\":\"other\",\"SWAT_3207\":true,\"common_offer_event_id\":\"75545e06-0817-4032-9d59-df34e0ace02e\",\"monetary_advertiser_amount\":-29.75,\"monetary_publisher_amount\":11.231,\"monetary_tapjoy_amount\":18.519,\"currency_reward\":11,\"spend_share\":0.378,\"reward_key\":\"d1e86a2d-1917-42fa-9afe-f5b4fff1d7e8\",\"multi_reward_event_name\":\"HIRE_HEROES3\",\"multi_reward_event_value\":\"N/A\",\"tjcid\":\"n0sToz78ApwWlBBRM_HaMV00gr3dOh7z0cubCY7UXzxO-70auqZv1HuXR3ssmoueyigrJWFMn-eO5ylGc1vOVZs1CMH4Jut8Ui5fIqdEw7IJsDbYY5IQcrWlVEyAsFEz\",\"batch\":false,\"currency_id\":\"f1f616e8-c3e3-433a-810f-5cc63da9697b\",\"time_ctj_pushed\":\"1715710447.2223256\"},\"time_ctj_pushed\":\"1715710447.2223256\",\"time_ctj_recieved\":\"1715710448.5653222\",\"click_source\":\"offerwall\",\"click_origin_source\":\"offerwall\",\"time_ctj_converted\":\"1715710448.769513\",\"is_admin_device\":false}]}",
       attributes={},
       md5_of_message_attributes=nil,
       message_attributes={}>,
     @queue_url="https://sqs.us-east-1.amazonaws.com/266171351246/v2_SendCurrency_Failures",
     @receipt_handle=
      "AQEBSHcv1W+f5pvWzqPcKxqNrAvCpZJp3dY58Gzdi2Eq9zhJj01z+l/ArgXK7DW2M42vbRzfw9/Xz+oHLz69TLxrzifNlIHcvjhVGHtz0jQ/K1RZu0+6RjcQ1FABVy7rs1l4dHjRYBVapuL0gXZ/efO8f6+xxKDJpx1K1QPEcWeuooAStq0dTAtlbWmY6M1tESv9Iqb2cmk7HQgpUAtKovvA+LTTqERvPC2jh9JrRLns+L6JI9Q39xZMldSdjHjdfXfYWKpJlGX5QxDjcSrSpYszXMXuOTXayUXsEcckcISTULMbOJPvC78GsOAP/MpHEVCQMXdtfHZUaoqL0n3ZPxIXsk1Nj9E0NhadU0Qr0F1lhZ/2nN6irEj/mZqfy2rpZL4mUlEcQuIVQEqcAi9zGpZ2OA==",
     @waiter_block_warned=false>]],
 @limit=nil,
 @size=1>```

Reply:
@mannychu I think the discussion above points you in the right direction. Please let us know if you need anything.

Reply:
@alanbrent @obrie At a very high level I think I understand what's going on, consumers receive messages from SQS, batch them, then send them to workers to get processed. However, I'm assuming we _can't_ just use the timestamp of when we send the batch to the workers, so we'd need to get the SentTimestamp for each individual message as we receive it. I believe it is possible to get the SentTimestamp (SQS message attribute) per message when we poll for messages, which is nice. 

I'm still trying to understand the flow and how things are passed around in Chore. I think once I understand I can figure out how to plug in this SentTimestamp and use it in the adjusted dedupe TTL. 

Couple questions here:

What method is being called here for `block.call`? I know it's a code block that's being passed around, just not familiar from which class/method.
https://github.com/Tapjoy/chore/blob/8959c633fb1ec00b8057b2cf029ef3af1772c8e9/lib/chore/queues/sqs/consumer.rb#L93C19-L95

In `worker.rb`, what is the`item` variable? Is it a unit of work, or a message from the SQS?
https://github.com/Tapjoy/chore/blob/8959c633fb1ec00b8057b2cf029ef3af1772c8e9/lib/chore/worker.rb#L45-L46

Reply:
I think one other thing is, is there a way to test this end to end?

Reply:
@obrie sorry I'm currently working on another prioritized task so wasn't able to figure out how to test this e2e. It doesn't look like there's a way to specifically test these changes through TIAB. Should I merge this Chore PR/version into master and update the TJS code to use the updated Chore version?

cc'ing @normchan-tj since he's looking into how to test Chore locally as well.

Reply:
@mannychu You can do the following without needing to merge or cut a release, to facilitate testing in TJS

1. Build the gem `gem build chore-core.gemspec`
1. Copy it to the `vendor/cache` in TJS
2. Update the reference in the TJS Gemfile
1. Run a `make dev` and once TJS comes up you should see that the `Gemfile.lock` file was updated with the new version you added

Once you have validated everything we can talk about cutting an actual release.

Reply:
@mannychu I see you haven't updated the version yet. Make sure to do so first before doing the above! I think 4.4.0 is probably appropriate, though perhaps 5.0 might be in order. @obrie Thoughts?